README.md
# Align-Memory

A simple C library for easy memory alignment.

## Table of Contents

- [Align-Memory](#align-memory)
  - [Table of Contents](#table-of-contents)
  - [Description](#description)
  - [Installation](#installation)
  - [Roadmap](#roadmap)
  - [Contributing](#contributing)
  - [Naming Convension](#naming-convension)
  - [License](#license)
  - [Project Status](#project-status)
  - [Acknowledgment](#acknowledgment)
  - [Contact](#contact)
  - [Versioning](#versioning)
  - [Changelog](#changelog)

## Description

This is a simple dynamic memory-alignment C library. It contains convenience functions for *getting aligned-memory
allocation size*, *aligned-memory location*, and *actual memory location from aligned-memory*.

See the list of features and functions implemented till now in [Project Status](#project-status).

## Installation

To get started with this project, download and install the following.

- Download and install **git**
  - If you use *Windows*, then go to [this link](https://git-scm.com/downloads) and download and install the suitable
    version.
  - If you use any stable version of *Debian/Ubuntu* then run this command in your terminal

    ```sh
    sudo apt-get install git
    ```

  - If you use *macOS* then install [homebrew](https://brew.sh/) if you don't already have it, then run this command
    in your terminal

    ```sh
    brew install git
    ```

- Run the command to clone this repository

  ```sh
  git clone https://github.com/KumarjitDas/Align-Memory.git
  ```

- Download and install a **C** compiler (*clang* or *gcc*)
  - If you use *Windows 10* then download and install a suitable version of **clang** from
    [this](https://releases.llvm.org/download.html) link. For **gcc**, use the suitable *MinGW* version from
    [this](http://mingw-w64.org/doku.php/download) link.
  - If you use any stable version of *Debian/Ubuntu* then run these commands in your terminal to download and install
    **clang** and **gcc** compilers

    ```sh
    sudo apt install clang
    ```

    ```sh
    sudo apt install gcc
    ```

  - In *macOS*, **clang** is the default **C** compiler. To download and install gcc, run this command in your terminal

    ```sh
    brew install gcc
    ```

- Download and install **Ninja** build system
  - If you use *Windows 10* then download and install **Ninja** from
    [here](https://github.com/ninja-build/ninja/releases).
  - If you use any stable version of *Debian/Ubuntu* then run this command in your terminal

    ```sh
    sudo apt-get install ninja-build
    ```

  - If you use *macOS* then run this command in your terminal

    ```sh
    brew install ninja
    ```

- Download and install **CMake** build automation tool
  - If you use *Windows 10* then download and install **CMake** from [here](https://cmake.org/download/).
  - If you use any stable version of *Debian/Ubuntu* then run this command in your terminal

    ```sh
    sudo apt-get -y install cmake
    ```

  - If you use *macOS* then run this command in your terminal

    ```sh
    brew install cmake
    ```

## Roadmap

See the [open issues](https://github.com/KumarjitDas/Align-Memory/issues) for a list of proposed
features/functionalities (and known issues).

The list of features and functions implemented till now is given in [Project Status](#project-status).

## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any
contributions you make are greatly appreciated.

- Fork this Project
- Create your *Feature Branch*

  ```sh
  git checkout -b feature/AmazingFeature
  ```

- Commit your *Changes*

  ```sh
  git commit -m 'Add some AmazingFeature'
  ```

- Push to the Branch

  ```sh
  git push origin feature/AmazingFeature
  ```

- Create a *Pull Request*

## Naming Convension

The naming convension used in this project will be listed soon.

## License

This project is distributed under the **MIT License**. See [LICENSE](LICENSE) for more information.

## Project Status

List of functionalities/features implemented so far.

- *kdi_get_size_for_memory_alignment*
- *kdi_align_memory*
- *kdi_get_actual_memory_from_aligned_memory*

## Acknowledgment

I appreciate these websites which helped me to make such good **README** file, and helped me to learn about project
versioning and keeping **CHANGELOG**.

- [Make a README](https://www.makeareadme.com/)
- [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
- [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## Contact

Twitter: [@kumarjitdas1999](https://twitter.com/kumarjitdas1999)

LinkedIn:
[Kumarjit Das | কুমারজিৎ দাস](https://www.linkedin.com/in/kumarjit-das/)

E-mail: [kumarjitdas1999@gmail.com](kumarjitdas1999@gmail.com)

Project link:
[GitHub-KumarjitDas-Align-Memory](https://github.com/KumarjitDas/Align-Memory)

## Versioning

This project uses **MAJOR**, **MINOR**, and **PATCH** version numbers for
versioning (v*MAJOR.MINOR.PATCH*).

- **MAJOR** version number indicates *new changes which may be incompatible with older versions*.
- **MINOR** version number indicates *addition of backwards-compatible features*.
- **PATCH** version number indicates *backwards-compatible bug fixes*, or minor mistake fixes like *spelling*,
  *character cases*, *punctuations*, and *indentation*.

## Changelog

The [Changelog](CHANGELOG.md) file contains all the information about the changes made to this project till now.


CHANGELOG.md
# Align-Memory Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.1/), and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.1.html).

## [Unreleased](https://github.com/KumarjitDas/Align-Memory/compare/v1.0.0...HEAD)

## [1.0.0](https://github.com/KumarjitDas/Align-Memory/compare/v0.9.0...v1.0.0) - 2022-01-17

### Added

- `KDI_BUILD_LIBRARY_TYPE` and `KDI_BUILD_LIBRARY_TYPE_LOWER` variables
- build types validator for **Ninja Multi-Config**
- more variables: `KDI_LIBRARY_OUTPUT_DIR`, `KDI_RUNTIME_OUTPUT_DIR`, `KDI_CURRENT_USER_PATH`, and
  `KDI_BUILD_DIRECTORY`
- build postfixes to [Examples](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Examples/CMakeLists.txt) and
  [Tests](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Tests/CMakeLists.txt) output names.

### Changed

- Using underscore and KDI namespace for local variable names
- `STATIC_POSTFIX` to `KDI_STATIC_POSTFIX`

## [0.9.0](https://github.com/KumarjitDas/Align-Memory/compare/v0.8.0...v0.9.0) - 2021-12-10

### Added

- `kdi_get_aligned_memory_backward` public API function
  [declaration/prototype](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Sources/include/align_memory/align_memory.h)
  and [definition](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Sources/src/align_memory/align_memory.c)
- Tests for `kdi_get_aligned_memory_backward` function in
  [kdi_get_aligned_memory_backward.c](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Tests/kdi_get_aligned_memory_backward.c) file
- Documentations for `kdi_get_aligned_memory_forward` and `kdi_get_aligned_memory_backward` functions
- Doxygen style documentations
- [example-002.c](https://github.com/KumarjitDas/Align-Memory/blob/v0.9.0/Examples/example-002.c) file

### Removed

- Old style documentation

## [0.8.0](https://github.com/KumarjitDas/Align-Memory/compare/v0.7.0...v0.8.0) - 2021-12-10

### Added

- `KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT` architecture macro in public header and all source files
- `NOTEs` for the users of this library in the function documentations
- `kdi_get_aligned_memory_forward` public API function
  [declaration/prototype](https://github.com/KumarjitDas/Align-Memory/blob/v0.8.0/Sources/include/align_memory/align_memory.h)
  and [definition](https://github.com/KumarjitDas/Align-Memory/blob/v0.8.0/Sources/src/align_memory/align_memory.c)
- Tests for `kdi_get_aligned_memory_forward` function in
  [kdi_get_aligned_memory_forward.c](https://github.com/KumarjitDas/Align-Memory/blob/v0.8.0/Tests/kdi_get_aligned_memory_forward.c) file

### Changed

- Generic variable names with appropriate 64 bit names in all source files

### Removed

- Unnecessary type casting in sources.
- `show_usage` function from
  [example-001.c](https://github.com/KumarjitDas/Align-Memory/blob/v0.8.0/Examples/example-001.c) file
- Implementation of memory alignment in `kdi_align_memory` function, using `kdi_get_aligned_memory_forward` api
  function instead

## [0.7.0](https://github.com/KumarjitDas/Align-Memory/compare/v0.6.0...v0.7.0) - 2021-12-10

### Added

- Documentation explaining explaining how memory alignment happens in
  [align_memory.c](https://github.com/KumarjitDas/Align-Memory/blob/v0.7.0/Sources/src/align_memory/align_memory.c)
  file

### Changed

- Public function names
  - `kdi_align_memory_get_allocation_size` to `kdi_get_size_for_memory_alignment`
  - `kdi_align_memory_get_aligned_address` to `kdi_align_memory`
  - `kdi_align_memory_get_actual_address` to `kdi_get_actual_memory_from_aligned_memory`

### Removed

- Triple underscores from private names
- Unnecessary comments in private source files

## [0.6.0](https://github.com/KumarjitDas/Align-Memory/compare/v0.5.0...v0.6.0) - 2021-12-05

### Added

- Documentation in all source files.
- License text as comments at the top of root
  [CMake](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/CMakeLists.txt) and
  [setup_project.cmake](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/CMake/setup_project.cmake) files and
  all the source files
- `write_status` function in to print configuring status in the terminal
- [config-static.cmake.in](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Configuration/config-static.cmake.in)
  file for static library configuration
- [config-shared.cmake.in](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Configuration/config-shared.cmake.in)
  file for shared library configuration
- More build scripts(windows batch files) in [Scripts](https://github.com/KumarjitDas/Align-Memory/tree/v0.6.0/Scripts)
  directory
  - [configure.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/configure.bat)
  - [configure-all.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/configure-all.bat)
  - [build.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/build.bat)
  - [build-all.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/build-all.bat)
  - [build-configs.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/build-configs.bat)
  - [build-all-configs.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/build-all-configs.bat)
  - [test.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/test.bat)
  - [test-all.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/test-all.bat)
  - [test-configs.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/test-configs.bat)
  - [test-all-configs.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/test-all-configs.bat)
  - [clean.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/clean.bat)
  - [clean-all.bat](https://github.com/KumarjitDas/Align-Memory/blob/v0.6.0/Scripts/clean-all.bat)

Tests/kdi_align_memory.c
/**
 * \file kdi_align_memory.c
 * \author Kumarjit Das (kumarjitdas1999@gmail.com)
 * \brief Tests for `kdi_align_memory` function.
 * \version 0.9.0
 * \date 2021-12-10
 *
 * \copyright Copyright (c) 2021
 *
 * License(MIT License):
 *
 * Copyright (c) 2021 Kumarjit Das | কুমারজিৎ দাস
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include "align_memory/align_memory.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <assert.h>

// #undef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

#ifdef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

int main() {
    uint64_t u64Usable_size, u64Alignment_size, u64Allocation_size;
    uint32_t *p32Memory, *p32Memory_aligned;

    // Test 1
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 1;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 2
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 2;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 3
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 4;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 4
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 8;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 5
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 16;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 6
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 32;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 7
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 64;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 8
    u64Usable_size = 1000 * sizeof(*p32Memory);
    u64Alignment_size = 128;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_align_memory(p32Memory,
                                         u64Alignment_size);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    return EXIT_SUCCESS;
}

#else  /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT not defined */
#endif /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT */


Tests/kdi_get_aligned_memory_forward.c
/**
 * \file kdi_get_aligned_memory_forward.c
 * \author Kumarjit Das (kumarjitdas1999@gmail.com)
 * \brief Tests for `kdi_get_aligned_memory_forward` function.
 * \version 0.9.0
 * \date 2021-12-10
 *
 * \copyright Copyright (c) 2021
 *
 * License(MIT License):
 *
 * Copyright (c) 2021 Kumarjit Das | কুমারজিৎ দাস
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include "align_memory/align_memory.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <assert.h>

// #undef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

#ifdef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

int main() {
    uint64_t u64Usable_size, u64Alignment_size, u64Allocation_size;
    uint32_t *p32Memory, *p32Memory_aligned;

    // Test 1
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 1;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 2
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 2;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 3
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 4;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 4
    u64Usable_size = 10 * sizeof(*p32Memory);
    u64Alignment_size = 8;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 5
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 16;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 6
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 32;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 7
    u64Usable_size = 100 * sizeof(*p32Memory);
    u64Alignment_size = 64;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    // Test 8
    u64Usable_size = 1000 * sizeof(*p32Memory);
    u64Alignment_size = 128;
    u64Allocation_size = kdi_get_size_for_memory_alignment(u64Usable_size,
                                                           u64Alignment_size);
    p32Memory = malloc(u64Allocation_size);
    if (p32Memory == NULL) {
        return EXIT_FAILURE;
    }
    p32Memory_aligned = kdi_get_aligned_memory_forward(p32Memory,
                                                       u64Alignment_size);
    assert((uint64_t)p32Memory_aligned >= (uint64_t)p32Memory);
    assert(((uint64_t)p32Memory_aligned % u64Alignment_size) == 0);
    free(p32Memory);

    return EXIT_SUCCESS;
}

#else  /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT not defined */
#endif /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT */


Examples/example-002.c
/**
 * \file example-002.c
 * \author Kumarjit Das (kumarjitdas1999@gmail.com)
 * \brief An example for `Align-Memory` library functions.
 * \version 0.9.0
 * \date 2021-12-10
 *
 * \copyright Copyright (c) 2021
 *
 * License(MIT License):
 *
 * Copyright (c) 2021 Kumarjit Das | কুমারজিৎ দাস
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include "align_memory/align_memory.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

// #undef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

#ifdef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

int main(int argc, char **argv) {
    if (argc < 2) {
        printf(
            "Usage: example-002 <align>\n"
            "    align - a positive integer between 1 and 128(including) and\n"
            "            it should be a power of 2\n");
        return EXIT_FAILURE;
    }

    printf("\n Align-Memory library example 2:\n\n");

    uint64_t u64Alignment_size = strtoull(argv[1], NULL, 10);
    int32_t *p32Array = malloc(1);
    if (p32Array == NULL) {
        fprintf(stderr, "ERROR: Could not allocate memory!\n");
        return EXIT_FAILURE;
    }
    int32_t *p32Forward_aligned_array = kdi_get_aligned_memory_forward(
        p32Array,
        u64Alignment_size);
    int32_t *p32Backward_aligned_array = kdi_get_aligned_memory_backward(
        p32Array,
        u64Alignment_size);
    printf(
        " Alignment size:                 %llu\n"
        " Array address:                  [%p]\n"
        " Forward aligned array address:  [%p]\n"
        " Backward aligned array address: [%p]\n",
        u64Alignment_size,
        p32Array,
        p32Forward_aligned_array,
        p32Backward_aligned_array);

    free(p32Array);
    return EXIT_SUCCESS;
}

#else  /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT not defined */
#endif /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT */


Examples/example-001.c
/**
 * \file example-001.c
 * \author Kumarjit Das (kumarjitdas1999@gmail.com)
 * \brief An example for `Align-Memory` library functions.
 * \version 0.9.0
 * \date 2021-12-10
 *
 * \copyright Copyright (c) 2021
 *
 * License(MIT License):
 *
 * Copyright (c) 2021 Kumarjit Das | কুমারজিৎ দাস
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include "align_memory/align_memory.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

// #undef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

#ifdef KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT

int main(int argc, char **argv) {
    if (argc < 3) {
        printf(
            "Usage: example-001 length align\n"
            "    length - a positive integer\n"
            "    align  - a positive integer between 1 and 128(including) "
            "and\n"
            "             it should be a power of 2\n");
        return EXIT_FAILURE;
    }

    printf("\n Align-Memory library example 1:\n\n");

    uint64_t u64Usable_length = strtoull(argv[1], NULL, 10);
    uint64_t u64Usable_size = u64Usable_length * sizeof(int32_t);
    uint64_t u64Alignment_size = strtoull(argv[2], NULL, 10);
    uint64_t u64Allocation_size = kdi_get_size_for_memory_alignment(
        u64Usable_size,
        u64Alignment_size);
    int32_t *p32Array = malloc(u64Allocation_size);
    if (p32Array == NULL) {
        fprintf(stderr, "ERROR: Could not allocate memory!\n");
        return EXIT_FAILURE;
    }
    int32_t *p32Aligned_array = kdi_align_memory(p32Array,
                                                 u64Alignment_size);
    int32_t *p32Actual_array = kdi_get_actual_memory_from_aligned_memory(
        p32Aligned_array);

    printf(
        " Usable length:         %llu\n"
        " Usable size:           %llu\n"
        " Alignment size:        %llu\n"
        " Allocation size:       %llu\n"
        " Array address:         [%p]\n"
        " Aligned array address: [%p]\n"
        " Actual array address:  [%p]\n",
        u64Usable_length,
        u64Usable_size,
        u64Alignment_size,
        u64Allocation_size,
        p32Array,
        p32Aligned_array,
        p32Actual_array);

    free(p32Array);
    return EXIT_SUCCESS;
}

#else  /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT not defined */
#endif /* KDI_ALIGN_MEMORY_ARCHITECTURE_64_BIT */
